File: solutions/python/math-geometry/multiply-strings.py
# LeetCode 43 — Multiply Strings (Medium)
# Category: Math & Geometry · Approach: Grade-School
# Time: O(m·n) | Space: O(m + n)
# Source: https://leetcode.com/problems/multiply-strings/

def multiply(num1: str, num2: str) -> str:
    if num1 == "0" or num2 == "0":
        return "0"
    m, n = len(num1), len(num2)
    pos = [0] * (m + n)
    for i in range(m - 1, -1, -1):
        for j in range(n - 1, -1, -1):
            mul = (ord(num1[i]) - 48) * (ord(num2[j]) - 48)
            p1, p2 = i + j, i + j + 1
            total = mul + pos[p2]
            pos[p2] = total % 10
            pos[p1] += total // 10
    res = "".join(map(str, pos)).lstrip("0")
    return res or "0"

File: solutions/typescript/math-geometry/multiply-strings.ts
// LeetCode 43 — Multiply Strings (Medium)
// Category: Math & Geometry · Approach: Grade-School
// Time: O(m·n) | Space: O(m + n)
// Source: https://leetcode.com/problems/multiply-strings/

function multiply(num1: string, num2: string): string {
  if (num1 === "0" || num2 === "0") return "0";
  const m = num1.length, n = num2.length;
  const pos = new Array(m + n).fill(0);
  for (let i = m - 1; i >= 0; i--) {
    for (let j = n - 1; j >= 0; j--) {
      const mul = (num1.charCodeAt(i) - 48) * (num2.charCodeAt(j) - 48);
      const p1 = i + j, p2 = i + j + 1;
      const sum = mul + pos[p2];
      pos[p2] = sum % 10;
      pos[p1] += Math.floor(sum / 10);
    }
  }
  let res = pos.join("").replace(/^0+/, "");
  return res === "" ? "0" : res;
}
